File: src_simple/python/trapeze.py
#!/usr/bin/env python
# -*- coding=utf-8 -*-

""" trapeze  -  11.03.2014 

Intégration par la méthode des trapèzes. L'intégration analytique 
de la fonction (x^2 - 3x - 6) e^-x entre 1 et 3 donne comme 
application numérique -2.525369

voir également : scipy.integrate.trapz
"""

from math import exp

def trapeze(a, b, npas):
    """ integration par la methode des trapeze """

    # initialisation
    integrale = 0.
    pas = (b - a) / float(npas)

    # calcul de l'integrale
    for i in range(npas):
        x = a + float(i) * pas
        integrale += pas * (f(x) + f(x + pas)) / 2.

    return integrale

def f(x):
    """ fonction numerique utilisée """
    return (x**2 - 3. * x - 6.) * exp(-x)

if __name__ == "__main__":
    # lecture de l'intervalle
    a = float(raw_input("entrer a : "))
    print("a = {0}".format(a))

    b = float(raw_input("entrer b : "))
    print("b = {0}".format(b))

    # nombre de segments
    npas = int(raw_input("entrer le nombre de pas : "))
    print("npas = {0}".format(npas))

    integrale = trapeze(a, b, npas)

    analytique = -2.525369
    print("Résultat  = {0}".format(integrale))
    print("Résidu    = {0}".format(analytique - integrale))
    print("Précision = {0}".format((analytique - integrale) / analytique * 100))

File: src_simple/f90/trapeze.f90
! trapeze  -  11.03.2014

! *******************************************************************
! Intégration par la méthode des trapèzes. L'intégration analytique 
! de la fonction (x^2 - 3x - 6) e^-x entre 1 et 3 donne comme 
! application numérique -2.525369
!********************************************************************

program trapeze

    implicit none

    integer         :: i, npas
    real            :: a, b, pas, integrale, x

    ! valeur exacte
    real, parameter :: analytique = -2.525369

    ! Lecture de l'intervalle
    write(*, *) "a = "
    read(*, *) a
    write(*, *) "a = ", a

    write(*, *) "b = "
    read(*, *) b
    write(*, *) "b = ", b

    ! Nombre de segments
    write(*, *) "npas = "
    read(*, *) npas
    write(*, *) "npas = ", npas

    ! initialisation
    integrale = 0.0
    pas = (b - a) / real(npas)

    ! calcul de l'integrale
    do i = 1, npas
        x = a + (i - 1) * pas
        integrale = integrale + pas * (f(x) + f(x + pas)) / 2.0
    end do

    ! résultats
    write(*, *) "Résultat  = ", integrale
    write(*, *) "Résidu    = ", analytique - integrale
    write(*, *) "Précision = ", (analytique - integrale) / analytique * 100.0

    contains
        
        ! fonction numérique utilisée
        real function f(x)
            implicit none
            real :: x
            f = (x**2 - 3 * x - 6) * exp(-x)
       end function f

end program trapeze
